### ISSUE TITLE
"gblinear" train-error become inf in r

### ISSUE THREAD

**Issue by 358463121** [dmlc/xgboost#2367]
For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.

## Environment info
Operating System: windows 10

Compiler:

Package used (python/R/jvm/C++):R

`xgboost` version used: 0.6-4

If you are using R package, please provide

1. The R `sessionInfo()`
```R
> sessionInfo()
R version 3.4.0 (2017-04-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=Chinese (Simplified)_China.936  LC_CTYPE=Chinese (Simplified)_China.936   
[3] LC_MONETARY=Chinese (Simplified)_China.936 LC_NUMERIC=C                              
[5] LC_TIME=Chinese (Simplified)_China.936    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] bnlearn_4.1.1 pcalg_2.4-5   self_0.1.0   

loaded via a namespace (and not attached):
 [1] igraph_1.0.1                graph_1.48.0                Rcpp_0.12.10               
 [4] cluster_2.0.6               magrittr_1.5                BiocGenerics_0.16.1        
 [7] sfsmisc_1.1-0               clue_0.3-51                 ggm_2.3                    
[10] lattice_0.20-35             tools_3.4.0                 parallel_3.4.0             
[13] grid_3.4.0                  data.table_1.10.4           corpcor_1.6.8              
[16] CompareCausalNetworks_0.1.5 abind_1.4-5                 xgboost_0.6-4              
[19] RBGL_1.46.0                 Matrix_1.2-9                gmp_0.5-12                 
[22] bdsmatrix_1.3-2             robustbase_0.92-6           fastICA_1.2-0              
[25] stringi_1.1.2               compiler_3.4.0              DEoptimR_1.0-6             
[28] stats4_3.4.0 
```
2. The command to install `xgboost` if you are not installing from source
install.packages("xgboost")
## Steps to reproduce

1.Download the data for linear regression [data.csv](https://github.com/358463121/markdown_notebook/raw/master/data.csv)
2.Read the data.csv by using the following code:
```R
data=data.table::fread("data.csv")
x=data[,1:5]
y=data[,6]
x=as.data.frame(x)
```
3.Just try  3-5 times to run the gblinear
```R
model<-xgboost::xgboost(data=as.matrix(x),label=as.numeric(y$y),
               verbose=1,nrounds=500,booster="gblinear",save_period=NULL)
```
And you would find something goes wrong.
Here is my result after 250 iterations.
```R
[250]	train-rmse:145373184131072.000000 
[251]	train-rmse:263565466402816.000000 
[252]	train-rmse:341494124773376.000000 
[253]	train-rmse:467568963878912.000000 
[254]	train-rmse:833250633187328.000000 
[255]	train-rmse:256638657232896.000000 
[256]	train-rmse:248258723053568.000000 
[257]	train-rmse:435057235853312.000000 
[258]	train-rmse:761494748790784.000000 
[259]	train-rmse:1287550933139456.000000 
[260]	train-rmse:2306050619342848.000000 
[261]	train-rmse:2908742173589504.000000 
[262]	train-rmse:4720008533901312.000000 
[263]	train-rmse:8235366251233280.000000 
[264]	train-rmse:13665744122281984.000000 
[265]	train-rmse:25019189521547264.000000 
[266]	train-rmse:47404361499803648.000000 
[267]	train-rmse:54305179943567360.000000 
[268]	train-rmse:86759224142462976.000000 
[269]	train-rmse:46830601113698304.000000 
[270]	train-rmse:68023064968822784.000000 
[271]	train-rmse:50070617067618304.000000 
[272]	train-rmse:75629598078926848.000000 
[273]	train-rmse:84318858084614144.000000 
[274]	train-rmse:111223847486750720.000000 
[275]	train-rmse:117351683486384128.000000 
[276]	train-rmse:116318546283200512.000000 
[277]	train-rmse:166796274710872064.000000 
[278]	train-rmse:211457732656496640.000000 
[279]	train-rmse:266076659961036800.000000 
[280]	train-rmse:Inf 
[281]	train-rmse:Inf 
[282]	train-rmse:Inf 
[283]	train-rmse:Inf 
```

## What have you tried?

1.I have tried to set a seed by using `set.seed(0)` but it did not work. I also tried the  `seed=0` param in xgboost, it did not work too. The result appeared randomly.
```R
model<-xgboost::xgboost(data=as.matrix(x),label=as.numeric(y$y),
               verbose=1,nrounds=500,booster="gblinear",save_period=NULL,seed=0) #seed do not work
```

**Comment by pommedeterresautee** [comment]
The error increases step after step, at some point R can't manage it and it becomes inf.
The issue is not inf, it is why your error explodes.

**Comment by 358463121** [comment]
@pommedeterresautee My issue is all about the billion RMSE error. I guess there is a bug in gblinear.

**Comment by khotilov** [comment]
I saw such behavior before. Setting the learning rate to a smaller value, e.g., down to at least 0.5 or 0.3 should help. Personally, I don't find the current eta=1 as a sensible default for gblinear, and I think it should be changed. Another option it to either reduce the number of threads or to run single-threaded. Note that you would also have to do the latter in order to get reproducible results since gblinear employs a "hogwild" (lock-free) parallelization in its cyclical gradient descent updates. While this free stochasticity might be beneficial in some situations, it sometimes may blow up when gradients are updated with high learning rates from several variables by several threads in an uncoordinated manner.
